src/db.c

```diff
@@ -1918,94 +1917,103 @@ int xreadGetKeys(struct redisCommand *cmd, robj **argv, int argc, getKeysResult
     } else {
-        raxRemove(server.cluster->slots_to_keys,indexed,keylen+2,NULL);
+        /* The removed entry was the first in the list. */
+        serverAssert(server.cluster->slots_to_keys[hashslot] == entry);
+        server.cluster->slots_to_keys[hashslot] = next;
     }
-    if (indexed != buf) zfree(indexed);
-}
-
-void slotToKeyAdd(sds key) {
-    slotToKeyUpdateKey(key,1);
 }
 
-void slotToKeyDel(sds key) {
-    slotToKeyUpdateKey(key,0);
-}
-
-/* Release the radix tree mapping Redis Cluster keys to slots. If 'async'
- * is true, we release it asynchronously. */
-void freeSlotsToKeysMap(rax *rt, int async) {
-    if (async) {
-        freeSlotsToKeysMapAsync(rt);
+/* Updates neighbour entries when an entry has been replaced (e.g. reallocated
+ * during active defrag). */
+void slotToEntryReplace(dictEntry *entry) {
+    dictEntry *next = dictEntryNextInSlot(entry);
+    dictEntry *prev = dictEntryPrevInSlot(entry);
+    if (next != NULL) {
+        dictEntryPrevInSlot(next) = entry;
+    }
+    if (prev != NULL) {
+        dictEntryNextInSlot(prev) = entry;
     } else {
-        raxFree(rt);
+        /* The replaced entry was the first in the list. */
+        sds key = entry->key;
+        unsigned int hashslot = keyHashSlot(key, sdslen(key));
+        server.cluster->slots_to_keys[hashslot] = entry;
     }
 }
 
-/* Empty the slots-keys map of Redis CLuster by creating a new empty one and
- * freeing the old one. */
-void slotToKeyFlush(int async) {
-    rax *old = server.cluster->slots_to_keys;
-
-    server.cluster->slots_to_keys = raxNew();
-    memset(server.cluster->slots_keys_count,0,
+/* Empty the slots-keys map of Redis Cluster. */
+void slotToKeyFlush(void) {
+    memset(server.cluster->slots_to_keys, 0,
+           sizeof(server.cluster->slots_to_keys));
+    memset(server.cluster->slots_keys_count, 0,
            sizeof(server.cluster->slots_keys_count));
-    freeSlotsToKeysMap(old, async);
 }
 
 /* Populate the specified array of objects with keys in the specified slot.
  * New objects are returned to represent keys, it's up to the caller to
  * decrement the reference count to release the keys names. */
 unsigned int getKeysInSlot(unsigned int hashslot, robj **keys, unsigned int count) {
-    raxIterator iter;
-    int j = 0;
-    unsigned char indexed[2];
-
-    indexed[0] = (hashslot >> 8) & 0xff;
-    indexed[1] = hashslot & 0xff;
-    raxStart(&iter,server.cluster->slots_to_keys);
-    raxSeek(&iter,">=",indexed,2);
-    while(count-- && raxNext(&iter)) {
-        if (iter.key[0] != indexed[0] || iter.key[1] != indexed[1]) break;
-        keys[j++] = createStringObject((char*)iter.key+2,iter.key_len-2);
-    }
-    raxStop(&iter);
+    unsigned int j = 0;
+    dictEntry *de = server.cluster->slots_to_keys[hashslot];
+    while (j < count && de != NULL) {
+        sds sdskey = dictGetKey(de);
+        keys[j++] = createStringObject(sdskey, sdslen(sdskey));
+        de = dictEntryNextInSlot(de);
+    }
```

Comment:
i think we should eliminate this function entirely, and just add the loop in the one place that calls it.
the current implementation is very wasteful since it copies the string to an `robj` and then copies it again to the output buffers (lost of excess malloc and memcpy calls).

note that in the distinct past the reply buffers where robj, so there was no excess memcpy, but IIRC starting with redis 4.0 (or maybe 3.2) we initially changed that to sds, and later to plain buffers.

Reply:
Done.